docs/API_DESIGN.md

```diff
@@ -0,0 +1,658 @@
+
+#### Example: Create & publish a `did:dht`
+
+```pseudocode!
+key_manager = new InMemoryKeyManager()
+public_key = key_manager.generate_key_material()
+did_dht = new DidDht(public_key)
+signer = key_manager.get_signer(public_key)
+did_dht.publish(signer)
+```
+
+#### Example: Create a `did:dht`, add to the Core Properties & publish
+
+> [!NOTE]
+> The call to the `new DidDht()` constructor only adds the minimum requirements to the DID Document.
+> If additional [Core Properties](https://www.w3.org/TR/did-core/#core-properties) are required, update the `document` data member prior-to the call to `publish()`.
+
+```pseudocode!
+key_manager = new InMemoryKeyManager()
+public_key = key_manager.generate_key_material()
+did_dht = new DidDht(public_key)
+
+/// Set the alsoKnownAs
+did_dht.document.alsoKnownAs = "did:example:efgh"
+/// Note: you could also add a verification method, set the controller etc.
+
+signer = key_manager.get_signer(public_key)
+did_dht.publish(signer)
+```
+
+#### Example: Instantiate an existing `did:dht`
+
+```pseudocode!
+uri = "did:dht:i9xkp8ddcbcg8jwq54ox699wuzxyifsqx4jru45zodqu453ksz6y"
+did_dht = new DidDht(uri)
+```
+
+#### Example: Update a `did:dht`
+
+```pseudocode!
+uri = "did:dht:i9xkp8ddcbcg8jwq54ox699wuzxyifsqx4jru45zodqu453ksz6y"
+did_dht = new DidDht(uri)
+
+/// Set the alsoKnownAs
+did_dht.document.alsoKnownAs = "did:example:efgh"
+/// Note: you could also add a verification method, set the controller etc.
+
+key_manager = new InMemoryKeyManager()
+public_key = key_manager.import_key(private_key) /// assume private_key pre-exists, eg. read from env var
+signer = key_manager.get_signer(public_key)
+
+did_dht.publish(signer)
+```
+
+#### Example: Resolve a `did:dht`
+
+```pseudocode!
+uri = "did:dht:i9xkp8ddcbcg8jwq54ox699wuzxyifsqx4jru45zodqu453ksz6y"
+resolution_result = DidDht.resolve(uri)
+```
+
+# Verifiable Credentials (VCs)
+
+## Data Model 1.1
+
+### `NamedIssuer`
+
+```pseudocode!
+CLASS NamedIssuer
+  PUBLIC DATA id: string
+  PUBLIC DATA name: string
+```
+
+### `VerifiableCredential`
+
+> [!WARNING]
+> The following is incomplete in that an `Object` is not currently supported in the Custom DSL; the matter of the `Object` below is a placeholder and expected to be completed in a subsequent version.
+
+> [!WARNING]
+> We need to consider default behaviors such as always including the base `@context` and `type`
+
+```pseudocode!
+CLASS VerifiableCredential
+  PUBLIC DATA @context: []string
+  PUBLIC DATA id: string
+  PUBLIC DATA type: []string
+  PUBLIC DATA issuer: string | NamedIssuer
+  PUBLIC DATA issuanceDate: string
+  PUBLIC DATA expirationDate: string?
+  PUBLIC DATA credentialSubject: Object  # 🚧 `Object` not supported 🚧
+  CONSTRUCTOR(context: []string, id: string, type: []string, issuer: string | NamedIssuer, issuanceDate: string, expirationDate: string?)
+  CONSTRUCTOR(vcjwt: string)
+  CONSTRUCTOR(vcjwt: string, verifier: Verifier)
+  METHOD sign(signer: Signer): string
```

Comment:
are you missing a `STATIC METHOD verify(vcjwt: string)` here?

Reply:
`CONSTRUCTOR(vcjwt: string)` executes cryptographic verification. We're in kind of a halfsy-state wherein I have some `STATIC METHOD`'s but I'm slightly skeptical to full-commit to the concept of a static method because the concept varies from language to language and IMO we misuse the concept in a number of our implementations.